src/pipewire/pipewire.cc

```diff
@@ -0,0 +1,575 @@
+
+    return pw_stream_new(m_core, _("Playback"), props);
+}
+
+bool PipeWireOutput::connect_stream(int format, int rate, int channels)
+{
+    m_rate = rate;
+    m_channels = channels;
+
+    uint8_t buffer[1024];
+    struct spa_pod_builder b = SPA_POD_BUILDER_INIT(buffer, sizeof buffer);
+
+    struct spa_audio_info_raw info = {
+        .format = to_pipewire_format(format),
+        .flags = SPA_AUDIO_FLAG_NONE,
+        .rate = m_rate,
+        .channels = m_channels
+    };
+
+    set_channel_map(&info, channels);
+
+    const struct spa_pod * params[1];
+    params[0] = spa_format_audio_raw_build(&b, SPA_PARAM_EnumFormat, &info);
+
+    auto stream_flags = static_cast<pw_stream_flags>(PW_STREAM_FLAG_AUTOCONNECT |
+                                                     PW_STREAM_FLAG_MAP_BUFFERS |
+                                                     PW_STREAM_FLAG_RT_PROCESS);
+
+    return pw_stream_connect(m_stream, PW_DIRECTION_OUTPUT, PW_ID_ANY,
+                             stream_flags, params, aud::n_elems(params)) == 0;
+}
+
+void PipeWireOutput::on_core_event_done(void * data, uint32_t id, int seq)
+{
+    PipeWireOutput * o = static_cast<PipeWireOutput *>(data);
+
+    if (id != PW_ID_CORE || seq != o->m_core_init_seq)
+        return;
+
+    spa_hook_remove(&o->m_registry_listener);
+    spa_hook_remove(&o->m_core_listener);
+
+    o->m_inited = true;
+    pw_thread_loop_signal(o->m_loop, false);
+}
+
+void PipeWireOutput::on_registry_event_global(void * data, uint32_t id, uint32_t permissions,
+                                              const char * type, uint32_t version,
+                                              const struct spa_dict * props)
+{
+    PipeWireOutput * o = static_cast<PipeWireOutput *>(data);
+
+    if (strcmp(type, PW_TYPE_INTERFACE_Node) != 0)
+        return;
+
+    const char * media_class = spa_dict_lookup(props, PW_KEY_MEDIA_CLASS);
+    if (!media_class)
+        return;
+
+    if (strcmp(media_class, "Audio/Sink") != 0)
+        return;
+
+    o->m_has_sinks = true;
+    o->m_core_init_seq = pw_core_sync(o->m_core, PW_ID_CORE, o->m_core_init_seq);
+}
+
+void PipeWireOutput::on_state_changed(void * data, enum pw_stream_state old,
+                                      enum pw_stream_state state, const char * error)
+{
+    PipeWireOutput * o = static_cast<PipeWireOutput *>(data);
+
+    if (o->m_ignore_state_change)
+        return;
+
+    if (state == PW_STREAM_STATE_UNCONNECTED ||
+        state == PW_STREAM_STATE_PAUSED ||
+        state == PW_STREAM_STATE_STREAMING)
+    {
+        pw_thread_loop_signal(o->m_loop, false);
+    }
+}
+
+void PipeWireOutput::on_process(void * data)
+{
+    PipeWireOutput * o = static_cast<PipeWireOutput *>(data);
+
+    if (!o->m_buffer_at)
+    {
+        pw_thread_loop_signal(o->m_loop, false);
+        return;
+    }
+
+    struct pw_buffer * b = pw_stream_dequeue_buffer(o->m_stream);
```

Comment:
Can `pw_stream_dequeue_buffer()` return `nullptr`?  The line below will crash if so:
```
struct spa_buffer * buf = b->buffer;
```